jetty-http/src/main/java/org/eclipse/jetty/http/HttpMethod.java

```diff
@@ -26,162 +26,179 @@
     {
-        for (HttpMethod method : HttpMethod.values())
-        {
-            CACHE.put(method.toString(), method);
-        }
+        return _type.ordinal() >= Type.IDEMPOTENT.ordinal();
     }
 
-    private final ByteBuffer _buffer;
-    private final byte[] _bytes;
+    public ByteBuffer asBuffer()
+    {
+        return _buffer.asReadOnlyBuffer();
+    }
 
-    HttpMethod()
+    public String asString()
     {
-        _bytes = StringUtil.getBytes(toString());
-        _buffer = ByteBuffer.wrap(_bytes);
+        return _method;
     }
 
-    public byte[] getBytes()
+    public String toString()
     {
-        return _bytes;
+        return _method;
     }
 
-    public boolean is(String s)
+    public static final Trie<HttpMethod> INSENSITIVE_CACHE = new ArrayTrie<>(252);
+    public static final Trie<HttpMethod> CACHE = new ArrayTernaryTrie<>(false, 300);
+    public static final Trie<HttpMethod> LOOK_AHEAD = new ArrayTernaryTrie<>(false, 330);
+    static
     {
-        return toString().equalsIgnoreCase(s);
+        for (HttpMethod method : HttpMethod.values())
+        {
+            if (!INSENSITIVE_CACHE.put(method.asString(), method))
+                throw new IllegalStateException("INSENSITIVE_CACHE too small: " + method);
+
+            if (!CACHE.put(method.asString(), method))
+                throw new IllegalStateException("CACHE too small: " + method);
+
+            if (!LOOK_AHEAD.put(method.asString() + ' ', method))
+                throw new IllegalStateException("LOOK_AHEAD too small: " + method);
+        }
     }
 
-    public ByteBuffer asBuffer()
+    /**
+     * Optimized lookup to find a method name and trailing space in a byte array.
+     *
+     * @param bytes Array containing ISO-8859-1 characters
+     * @param position The first valid index
+     * @param limit The first non valid index
+     * @return An HttpMethod if a match or null if no easy match.
+     */
+    public static HttpMethod lookAheadGet(byte[] bytes, final int position, int limit)
     {
-        return _buffer.asReadOnlyBuffer();
+        int len = limit - position;
+        if (limit > 3)
+        {
+            // Short cut for GET
+            if (bytes[position] == 'G' && bytes[position + 1] == 'E' && bytes[position + 2] == 'T' && bytes[position + 3] == ' ')
+                return GET;
+            // Otherwise lookup in the Trie
+            return LOOK_AHEAD.getBest(bytes, position, len);
+        }
+        return null;
     }
 
-    public String asString()
+    /**
+     * Optimized lookup to find a method name and trailing space in a byte array.
+     *
+     * @param buffer buffer containing ISO-8859-1 characters, it is not modified.
+     * @return An HttpMethod if a match or null if no easy match.
+     * @deprecated Not used
+     */
+    @Deprecated
+    public static HttpMethod lookAheadGet(ByteBuffer buffer)
     {
-        return toString();
+        return LOOK_AHEAD.getBest(buffer, 0, buffer.remaining());
```

Comment:
Is that extra `LOOK_AHEAD` trie worth it? Since it is just a copy of the `CACHE` one with the extra whitespace, wouldn't something like the following be as fast and avoid duplication?

```java
HttpMethod method = CACHE.getBest(buffer, 0, buffer.remaining());
if (method == null)
    return null;

int whitespaceIndex = method.name().length();
if (buffer.remaining() == whitespaceIndex)
    return null;

byte b = buffer.get(whitespaceIndex);
return b == ' ' ? method : null;
```

Reply:
That looks like the kind of thing your PR in 10 can resolve :)
Plus I kept it separate from CACHE as I wanted to allow you to remove the CACHE if it is not used for getBest.

Reply:
Also look at the impl I just did.  I think the buffer.getInt lookup could be nice and fast for the GET method, so could make it very worth it... but I'm guessing it is faster.